pyiron_base/generic/util.py

```diff
@@ -57,3 +61,119 @@ def static_isinstance(obj, obj_type):
         return obj_type in obj_class_lst
     else:
         raise TypeError()
+
+
+class Deprecator:
+    """
+    Decorator class to mark functions and methods as deprecated with a uniform
+    warning message at the time the function is called.  The message has the
+    form
+
+        {function_name} is deprecated: {message}. It is not guaranteed to be in
+        service after {version}.
+
+    unless `pending=True` was given.  Then the message will be
+
+        {function_name} will be deprecated in a future version: {message}.
+
+    If message and version are not initialized or given during the decorating
+    call the respective parts are left out from the message.
+
+    >>> deprecate = Deprecator()
+    >>> @deprecate
+    ... def foo(a, b):
+    ...     pass
+    >>> foo(1, 2)
+    DeprecationWarning: __main__.foo is deprecated
+
+    >>> @deprecate("use bar() instead")
+    ... def foo(a, b):
+    ...     pass
+    >>> foo(1, 2)
+    DeprecationWarning: __main__.foo is deprecated: use bar instead
+
+    >>> @deprecate("use bar() instead", version="0.4.0")
+    ... def foo(a, b):
+    ...     pass
+    >>> foo(1, 2)
+    DeprecationWarning: __main__.foo is deprecated: use bar instead.  It is not
+    guaranteed to be in service in vers. 0.4.0
+
+    >>> deprecate = Deprecator(message="pyiron says no!", version="0.5.0")
+    >>> @deprecate
+    ... def foo(a, b):
+    ...     pass
+    >>> foo(1, 2)
+    DeprecationWarning: __main__.foo is deprecated: pyiron says no!  It is not
+    guaranteed to be in service in vers. 0.5.0
+    """
+
+    def __init__(self, message=None, version=None, pending=False):
+        """
+        Initialize default values for deprecation message and version.
+
+        Args
+        ----
+            message (str):
+                default deprecation message
+            version (str):
+                default version after which the function will be removed
```

Comment:
The function **will** be removed or **might** be removed? I would indeed prefer to systematically remove all deprecated functions (in which case this DocString is correct), but since I'm a rather weak programmer and frankly have no idea how to automatise it, I could only suggest to change it to "might be"...

Either way, "will be removed" and "not guaranteed" (which appears in the warning) are not compatible.

Reply:
I took the warning message to be close to what our manual message say right now, so I'll just change the docstring.  As I said I'd want to explore automating it, but it's not so easy on the software side I think.   Worst case would be to grep "deprecate" + version number before each bump, which isn't terribly much work.